src/frequenz/sdk/actor/power_distributing/battery_pool.py

```diff
@@ -0,0 +1,121 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+"""Class that stores pool of batteries and manage them."""
+
+import asyncio
+import logging
+from asyncio import ALL_COMPLETED
+from typing import Set
+
+from ..._data_ingestion.battery_utils import BatteryUtils
+from .result import PartialFailure, Result, Success
+
+_logger = logging.getLogger(__name__)
+
+
+class BatteryPool:
+    """Holds bool of batteries and returns data from them."""
+
+    def __init__(
+        self,
+        battery_ids: Set[int],
+        max_data_age_sec: float,
+        max_blocking_duration_sec: float,
+    ) -> None:
+        """Create object instance.
+
+        Note:
+            The object instance is fully created after running `run` method.
```

Comment:
Another option is to have a `BatteryPool.new()` async class method that creates and returns a BatteryPool instance.  And in the class documentation say "use `pool = await BatteryPool.new(...)` to create".

``` python
from __future__ import annotations

class BatteryPool:
    """Use new() to create."""
    _CreatedThroughNew = object()

    def __init__(..., created_by: Any = None):
        """Not for direct use.  Use new() to create."""
        if created_by is not _CreatedThroughNew:
             raise RuntimeError(...)
        ... rest of init

    async def new(...) -> BatteryPool:
        await etc
        return BatteryPool(...., cls._CreatedThroughNew)

```

Reply:
Yes but then it will be impossible to create BatteryPool in any constructor.
So user would have to either:
1) In constructor define battery pool as: `battery_pool: Optional[BatteryPool] = None` -> Create battery pool in other(async) method -> before every use check None condition. (which would end up (in most cases) with ignoring pylint/mypy warnings)
2) Define BatteryPool outside constructor and ingore mypy/pylint warnings that object is defined outside `__init__`

So from usage perspective it is better to define it in this way :(
It is true , that user has to remember about `run` method. But at least it is easy to use.

Reply:
True,  okay. :-)

Then what do you think about the name `async_init`?  `run` sounds like it would start a long running task,  but that's not the case.

Reply:
I go with `init` because:
* `async_init` would look like: `async def async_init(self)` . 
* `init` is very close to `__init__` to so it makes sense it is part of the init step.
*   `init_batteries` - it is good, but can be misunderstood. User might wonder why he has to init batteries etc.
* BatteryPool is private class, now. We can rename `init` to anything else when we publish. And until we publish it,  it will be no harm to user :)

Reply:
The only issue is that now when I say "the init method", it is ambiguous which one I mean.

We'll probably end up calling it the "async init" anyway.

> `async_init` would look like: `async def async_init(self)` .

Is that bad?

Reply:
async before function definition tells that this method is async.

Reply:
Ok I will rename it to async_init.

Reply:
Done

Reply:
Damn GitHub, it lost my comment when I hit unresolve conversation :sob: 

I tried to say that I don't fully agree with not being able to use this class as other class attributes in constructors. There is another option, which is what happens in general with asyc. Async is viral, and if you want to use it, you need to make your stuff async and IMHO this should happen with constructors too, otherwise we are just making users need to write more error prone code.

So IMHO users that want to have a `BatteryPool` in their class should make their class *async-constructible* too, otherwise they will also need to split sync construction from async construction. We can add a utility class `AsyncConstructible` to allow users to also implement their own async constructible classes more easily, for example:

```py
class NotSyncConstructible(AssertError):
    pass


class AsyncConstructible(ABC):
    def __init__(self) -> None:
        raise NotSyncConstructible()

    @abstractmethod
    @classmethod
    async def async_new(cls) -> Self:  # or AsyncConstructible if it still doesn't work
        return cls.__new__(cls)


class BatteryPool(AsyncConstructible):
    @classmethod
    def async_new(cls, ...) -> Self:  # or BatteryPool if it still doesn't work
        self = await super().__async_new()
        # Assign attributes, do more async stuff
        return self
```

We just need to check if we can somehow document instance attributes.

All that said, I'm also fine to leave this PR as is and create a new issue for this, as it is not strictly related to the PR and it might need more discussion.

PS: I know this also add some error proneness, as users need to make sure to define their classes correctly, but at least you need to define a class (and get it right) only once, while if the error proneness is in the usage, you have to do it several times, having more chances to mess it up.

Reply:
```
class BatteryPool(AsyncConstructible):
    @classmethod
    def async_new(cls, ...) -> Self:  # or BatteryPool if it still doesn't work
        self = await super().__async_new()
        # Assign attributes, do more async stuff
        return self
```
About line     # Assign attributes, do more async stuff
If attribute was not assigned in method with self argument, then pylint doesn't recognize this attribute.  So there is lots of pylint errors that we have to quiet to make it work.

We should skip this topic in this PR and continue in other issue.